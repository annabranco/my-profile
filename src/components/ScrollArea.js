import React from 'react';
import InbetweenBar from './InbetweenBar';
import Developer from './Developer';
import Formation from './Formation';
import ScrollItem from './ScrollItem';
import '../styles/components/ScrollArea.css';


class ScrollArea extends React.Component {


  // determineBlocks() {
  //   let BlocksToRender = [];
  //
  //   for (let i = 0; i < 7; i++) {
  //     BlocksToRender.push(
  //
  //       <ScrollItem
  //           num={i}
  //           bg={`url('${this.props.bgs[i]}')`}
  //           handleAnimation={this.props.handleAnimation}
  //           title={this.props.blockContents[i][this.props.language].title}
  //       />
  //     ) ;
  //   }
  //   return BlocksToRender;
  // }


  render () {

    return (
      	<section className="main__images">


			<InbetweenBar title={this.props.texts.Developer[this.props.language].title} />
			<Developer
				  texts={this.props.texts}
				  language={this.props.language}
			/>

			<InbetweenBar texts={this.props.texts} />
			<Formation
				  texts={this.props.texts}
				  language={this.props.language}
			/>

      	</section>
    );
  }

}

export default ScrollArea;
