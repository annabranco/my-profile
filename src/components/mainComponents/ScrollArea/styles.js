import styled, { css } from 'styled-components';

export const ScrollAreaWrapper = styled.main`
  height: ${`${window.innerHeight}px`};
  overflow: hidden;
  position: absolute;

  ${props =>
    props.langModalVisible &&
    css`
      overflow-y: hidden;
    `}

  @media all and (min-width: 768px) {
    position: unset;
    overflow-y: scroll;
  }
`;
ScrollAreaWrapper.displayName = 'Scroll Area';
