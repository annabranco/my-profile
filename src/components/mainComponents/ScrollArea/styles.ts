import styled, { css } from 'styled-components';

interface ScrollAreaWrapperProps
  extends React.ComponentPropsWithoutRef<'main'> {
  langModalVisible: boolean;
  isSeabedElementOpened: boolean;
}

export const ScrollAreaWrapper = styled.main<ScrollAreaWrapperProps>`
  height: ${`${window.innerHeight}px`};
  overflow: hidden;
  position: absolute;
  scroll-snap-type: y proximity;

  ${({ langModalVisible }) =>
    langModalVisible &&
    css`
      overflow-y: hidden;
    `}

  @media all and (min-width: 768px) {
    position: unset;
    overflow-y: ${({ isSeabedElementOpened }) =>
      isSeabedElementOpened ? 'hidden' : 'scroll'};
  }
`;
ScrollAreaWrapper.displayName = 'Scroll Area';

interface ScrollSectionElementProps
  extends React.ComponentPropsWithoutRef<'div'> {
  last: boolean;
}

export const ScrollSectionElement = styled.div<ScrollSectionElementProps>`
  scroll-snap-align: ${({ last }) => last && 'end'};
`;
ScrollSectionElement.displayName = 'ScrollSectionElement';
