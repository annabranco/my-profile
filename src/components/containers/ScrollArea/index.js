import React, { Component, createRef } from 'react';
import { PropTypes } from 'prop-types';
import DeveloperProfile from '../DeveloperProfile';
import Seabed from '../Seabed';
import { InbetweenBar, Experiences, MyInfoPage } from '../../views';
import {
  textsPropType,
  projectsPropType,
  skillGroupsPropType,
  formationPropType,
  experiencesPropType
} from '../../../types';
import { ScrollAreaWrapper } from './styles';

class ScrollArea extends Component {
  static propTypes = {
    texts: textsPropType.isRequired,
    projects: PropTypes.arrayOf(projectsPropType).isRequired,
    skills: PropTypes.arrayOf(skillGroupsPropType).isRequired,
    experiences: PropTypes.arrayOf(experiencesPropType).isRequired,
    formation: PropTypes.arrayOf(formationPropType).isRequired,
    triggerThankYouMessage: PropTypes.func.isRequired,
    displayThanksMessage: PropTypes.bool.isRequired,
    languageModalIsVisible: PropTypes.bool.isRequired
  };

  scrollAreaRef = createRef();

  state = {
    cuePointsActivated: new Set()
  };

  SECTIONS_INTERVAL_POINTS = {
    developer: {
      scrollAreaStart: 350,
      scrollAreaEnd: 1540
    },
    experiences: {
      scrollAreaStart: 900,
      scrollAreaEnd: 1900
    }
  };

  SCROLL_CUEPOINTS = {};

  experiencesSectionIds = new Set();

  cuePointsActivated = new Set();

  componentDidMount() {
    this.calculateExperiencesScroll();
  }

  // handleAdjustExpandedProjectsView = adjust =>
  //   this.setState({
  //     adjustedSize: adjust
  //   });

  // toggleDynamicElements = (section, element, action) => {
  //   const key = `${section}Activation`;
  //   const result = action === SHOW_ACTION;
  //   this.setState(prevState => ({
  //     [key]: {
  //       ...prevState[key],
  //       [element]: result
  //     }
  //   }));
  // };

  calculateExperiencesScroll = () => {
    const { experiences } = this.props;
    const NUMBER_OF_EXPERIENCES_DISPLAYED = experiences.length;
    const {
      scrollAreaStart,
      scrollAreaEnd
    } = this.SECTIONS_INTERVAL_POINTS.experiences;
    const intervalWindowForEachExperience =
      (scrollAreaEnd - scrollAreaStart) / NUMBER_OF_EXPERIENCES_DISPLAYED;
    let index = 0;

    for (
      let position = scrollAreaStart;
      position < scrollAreaEnd;
      position += intervalWindowForEachExperience
    ) {
      this.SCROLL_CUEPOINTS = {
        ...this.SCROLL_CUEPOINTS,
        [position]: experiences[index].id
      };
      this.experiencesSectionIds.add(experiences[index].id);
      index += 1;
    }
    this.SCROLL_CUEPOINTS = {
      ...this.SCROLL_CUEPOINTS,
      [scrollAreaEnd]: 'experiencesSection'
    };
  };

  handleScroll = event => {
    const { cuePointsActivated } = this.state;
    const scrollPosition = event.target.scrollTop;
    const cuePointsActivatedBeforeScrolling = [...cuePointsActivated].join(' ');
    const cuePointsActivatedAfterScrolling = cuePointsActivated;

    console.log('$$$ scrollPosition', scrollPosition);

    Object.keys(this.SCROLL_CUEPOINTS).forEach(cuePoint => {
      if (scrollPosition < Number(cuePoint)) {
        cuePointsActivatedAfterScrolling.delete(
          this.SCROLL_CUEPOINTS[cuePoint]
        );
      } else if (scrollPosition >= Number(cuePoint)) {
        if (this.SCROLL_CUEPOINTS[cuePoint] === 'experiencesSection') {
          this.experiencesSectionIds.forEach(experienceCuePoint => {
            cuePointsActivatedAfterScrolling.delete(experienceCuePoint);
          });
        } else {
          cuePointsActivatedAfterScrolling.add(this.SCROLL_CUEPOINTS[cuePoint]);
        }
      }
    });

    // Checks if activated cuepoints have changed before updating state
    if (
      (cuePointsActivatedBeforeScrolling !== '' ||
        [...cuePointsActivatedAfterScrolling].join(' ') !== 0) &&
      cuePointsActivatedBeforeScrolling !==
        [...cuePointsActivatedAfterScrolling].join(' ')
    ) {
      this.setState({ cuePointsActivated: cuePointsActivatedAfterScrolling });
    }
    // if (scrollPosition) {
    //   // Toggle Developer section
    //   if (scrollPosition >= 350 && scrollPosition <= 1540) {
    //     this.toggleDynamicElements('developer', 'adalab', SHOW_ACTION);
    //   } else {
    //     this.toggleDynamicElements('developer', 'adalab', HIDE_ACTION);
    //   }
    //   if (scrollPosition >= 500 && scrollPosition <= 1540 + adjustedSize) {
    //     this.toggleDynamicElements('developer', 'projects', SHOW_ACTION);
    //   } else {
    //     this.toggleDynamicElements('developer', 'projects', HIDE_ACTION);
    //   }
    //   if (scrollPosition >= 450 && scrollPosition <= 1540 + adjustedSize) {
    //     this.toggleDynamicElements('developer', 'skills', SHOW_ACTION);
    //   } else {
    //     this.toggleDynamicElements('developer', 'skills', HIDE_ACTION);
    //   }
  };

  resetScrollPosition = delay =>
    setTimeout(() => this.scrollAreaRef.current.scrollTo(0, 0), delay);

  render() {
    const { cuePointsActivated } = this.state;
    const {
      texts,
      triggerThankYouMessage,
      displayThanksMessage,
      languageModalIsVisible,
      projects,
      skills,
      formation,
      experiences
    } = this.props;

    return (
      <ScrollAreaWrapper
        languageModalIsVisible={languageModalIsVisible}
        onScroll={this.handleScroll}
        ref={this.scrollAreaRef}
      >
        <MyInfoPage
          texts={texts.infoPage}
          displayThanksMessage={displayThanksMessage}
        />
        <InbetweenBar title={texts.sections.technical} />
        <DeveloperProfile
          texts={texts.developer}
          projects={projects}
          skills={skills}
          handleAdjustExpandedProjectsView={
            this.handleAdjustExpandedProjectsView
          }
          language={texts.languages.languageCode}
          cuePointsActivated={[...cuePointsActivated]}
        />

        <InbetweenBar title={texts.sections.experience} />
        <Experiences
          texts={texts.global}
          experiences={experiences}
          language={texts.languages.languageCode}
          cuePointsActivated={[...cuePointsActivated]}
        />

        <InbetweenBar title={texts.sections.other} />
        <Seabed
          texts={texts.seabed}
          globalTexts={texts.global}
          formation={formation}
          textsOtherSkills={texts.otherSkills}
          triggerThankYouMessage={triggerThankYouMessage}
          resetScrollPosition={this.resetScrollPosition}
        />
      </ScrollAreaWrapper>
    );
  }
}

export default ScrollArea;
