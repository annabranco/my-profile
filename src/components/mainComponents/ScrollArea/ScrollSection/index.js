import React from 'react';
import { node, string } from 'prop-types';
import { SectionContent, SectionHeader, Title } from './styles';
import { isDesktop } from '../../../../utils/device';

const ScrollSection = ({ children, title }) => (
  <section id={title}>
    {isDesktop && (
      <SectionHeader>
        <Title>{title}</Title>
      </SectionHeader>
    )}
    <SectionContent>{children}</SectionContent>
  </section>
);

ScrollSection.propTypes = {
  children: node.isRequired,
  title: string.isRequired
};

export default ScrollSection;
