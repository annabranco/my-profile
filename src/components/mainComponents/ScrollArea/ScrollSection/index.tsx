import { ReactElement, ReactNode } from 'react';
import { bool, node, string } from 'prop-types';
import { isDesktop } from '../../../../utils/device';
import { SectionContent, SectionHeader, Title } from './styles';
import { ScrollSectionElement } from '../styles';

interface Props {
  children: ReactNode | ReactNode[];
  datae2eid?: string;
  last: boolean;
  title: string;
}

const ScrollSection = ({
  children,
  datae2eid,
  last,
  title
}: Props): ReactElement => (
  <ScrollSectionElement id={title} last={last}>
    {isDesktop && (
      <SectionHeader>
        <Title data-e2e-id={datae2eid}>{title}</Title>
      </SectionHeader>
    )}
    <SectionContent>{children}</SectionContent>
  </ScrollSectionElement>
);

ScrollSection.propTypes = {
  children: node.isRequired,
  datae2eid: string,
  last: bool,
  title: string.isRequired
};
ScrollSection.defaultProps = {
  datae2eid: null,
  last: false
};
export default ScrollSection;
