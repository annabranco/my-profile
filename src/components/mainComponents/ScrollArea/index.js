import React, { useEffect, useRef } from 'react';
import { useSelector } from 'react-redux';
import { bool } from 'prop-types';
import Experiences from '../../sections/Experiences';
import MyInfoPage from '../../sections/MyInfoPage';
import Projects from '../../sections/Projects';
import ScrollSection from './ScrollSection';
import Skills from '../../sections/Skills';
import Seabed from '../../sections/Seabed';
import {
  experiencesSelector,
  secionsTextsSelector
} from '../../../redux/selectors';
import { useStateWithLabel } from '../../../utils/hooks';
import {
  ADJUST_SCROLL,
  CUEPOINTS,
  SKILLS_FIRST_ROW,
  SKILLS_SECOND_ROW,
  SKILLS_THIRD_ROW,
  SEABED_AREA,
  INFO_PAGE_SECTION,
  SKILLS_SECTION,
  PROJECTS_SECTION,
  EXPERIENCES_SECTION,
  SEABED_SECTION,
  ACTIVE_SECTION
} from '../../../constants';
import { ScrollAreaWrapper } from './styles';
import { isDesktop } from '../../../utils/device';

const SCROLL_CUEPOINTS = new Map();

const ScrollArea = ({ langModalVisible }) => {
  const experiences = useSelector(experiencesSelector);
  const sectionsTexts = useSelector(secionsTextsSelector);

  const [adjustScroll, changeAdjust] = useStateWithLabel(0, ADJUST_SCROLL);
  const [cuePointsActivated, updateCuepoints] = useStateWithLabel(
    new Set(),
    CUEPOINTS
  );
  const [activeSection, changeActiveSection] = useStateWithLabel(
    SKILLS_SECTION,
    ACTIVE_SECTION
  );
  const scrollAreaRef = useRef();

  const SECTIONS_INTERVAL_POINTS = {
    skills: {
      scrollAreaStart: 200,
      scrollAreaEnd: 1200
    },
    experiences: {
      scrollAreaStart: 1400 + adjustScroll,
      scrollAreaEnd: 2100 + adjustScroll
    },
    seabed: {
      scrollAreaStart: 2600 + adjustScroll
    }
  };

  const experiencesSectionIds = new Set();

  const handleScroll = event => {
    const scrollPosition = event.target.scrollTop;
    const cuePointsAfterScrolling = new Set(cuePointsActivated);

    // eslint-disable-next-line no-underscore-dangle
    if (window._debug) {
      console.debug('scrollPosition', scrollPosition); // eslint-disable-line no-console
      console.debug('cuePointsActivated', cuePointsActivated); // eslint-disable-line no-console
    }

    SCROLL_CUEPOINTS.forEach((cuePointName, cuePoint) => {
      if (scrollPosition < cuePoint) {
        cuePointsAfterScrolling.delete(cuePointName);
      } else if (scrollPosition >= Number(cuePoint)) {
        if (cuePointName === 'experiencesSection') {
          // Makes experiences also go out when scrolling down
          //     experiencesSectionIds.forEach(experienceCuePoint => {
          //       cuePointsAfterScrolling.delete(experienceCuePoint);
        } else {
          cuePointsAfterScrolling.add(cuePointName);
        }
      }
    });
    // Prevents updating if cuepoints haven't changed
    if (cuePointsActivated.size !== cuePointsAfterScrolling.size) {
      updateCuepoints(new Set(cuePointsAfterScrolling));
    }
  };

  const resetScrollPosition = delay =>
    setTimeout(() => scrollAreaRef.current.scrollTo(0, 0), delay);

  const adjustScrollAfterThumbnails = adjustment => {
    changeAdjust(adjustment);
  };

  useEffect(() => {
    const calculateSkillsScroll = () => {
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart,
        SKILLS_FIRST_ROW
      );
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart + 300,
        SKILLS_SECOND_ROW
      );

      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart + 400,
        SKILLS_THIRD_ROW
      );
    };

    const calculateExperiencesScroll = () => {
      const NUMBER_OF_EXPERIENCES_DISPLAYED = experiences.length;
      const {
        scrollAreaStart,
        scrollAreaEnd
      } = SECTIONS_INTERVAL_POINTS.experiences;
      const intervalWindowForEachExperience =
        (scrollAreaEnd - scrollAreaStart) / NUMBER_OF_EXPERIENCES_DISPLAYED;
      let index = 0;

      for (
        let position = scrollAreaStart;
        position < scrollAreaEnd;
        position += intervalWindowForEachExperience
      ) {
        SCROLL_CUEPOINTS.set(Math.floor(position), experiences[index].id);
        experiencesSectionIds.add(experiences[index].id);
        index += 1;
      }
      SCROLL_CUEPOINTS.set(scrollAreaEnd, 'experiencesSection');
    };

    const calculateSeabedScroll = () => {
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.seabed.scrollAreaStart,
        SEABED_AREA
      );
    };

    calculateSkillsScroll();
    calculateExperiencesScroll();
    calculateSeabedScroll();
  }, [SECTIONS_INTERVAL_POINTS, experiences, experiencesSectionIds]);

  useEffect(() => {
    console.log('$$$ activeSection', activeSection);
    if (!isDesktop && activeSection === SKILLS_SECTION) {
      const allCuePoints = [
        SKILLS_FIRST_ROW,
        SKILLS_SECOND_ROW,
        SKILLS_THIRD_ROW
      ];
      updateCuepoints(new Set(allCuePoints));
    }
  }, [isDesktop, activeSection]);

  return (
    <ScrollAreaWrapper
      langModalVisible={langModalVisible}
      onScroll={handleScroll}
      ref={scrollAreaRef}
    >
      {(isDesktop || activeSection === INFO_PAGE_SECTION) && (
        <MyInfoPage changeActiveSection={changeActiveSection} />
      )}

      {(isDesktop || activeSection === SKILLS_SECTION) && (
        <ScrollSection title={sectionsTexts.technical}>
          <Skills
            cuePointsActivated={cuePointsActivated}
            changeActiveSection={changeActiveSection}
          />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === PROJECTS_SECTION) && (
        <ScrollSection title={sectionsTexts.projects}>
          <Projects
            adjustScrollAfterThumbnails={adjustScrollAfterThumbnails}
            changeActiveSection={changeActiveSection}
          />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === EXPERIENCES_SECTION) && (
        <ScrollSection title={sectionsTexts.experience}>
          <Experiences
            changeActiveSection={changeActiveSection}
            cuePointsActivated={cuePointsActivated}
          />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === SEABED_SECTION) && (
        <ScrollSection title={sectionsTexts.other}>
          <Seabed
            changeActiveSection={changeActiveSection}
            cuePointsActivated={cuePointsActivated}
            resetScrollPosition={resetScrollPosition}
          />
        </ScrollSection>
      )}
    </ScrollAreaWrapper>
  );
};

ScrollArea.propTypes = {
  langModalVisible: bool.isRequired
};

export default ScrollArea;
