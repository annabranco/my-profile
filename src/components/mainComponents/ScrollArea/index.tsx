import React, {
  Dispatch,
  ReactElement,
  SetStateAction,
  useCallback,
  useEffect,
  useRef
} from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { bool, func } from 'prop-types';
import { IExperiencesType, ISectionsText } from '../../../types/interfaces';
import {
  currentSecionSelector,
  experiencesSelector,
  sectionsTextsSelector
} from '../../../redux/selectors';
import { changeSection } from '../../../redux/actions/sections';
import { isDesktop } from '../../../utils/device';
import { useStateWithLabel } from '../../../utils/hooks';
import { getNextSection, getPreviousSection } from '../../../utils/sections';
import SwipeInstructions from '../../elements/SwipeInstructions';
import Experiences from '../../sections/Experiences';
import Formation from '../../sections/Formation';
import MyInfoPage from '../../sections/MyInfoPage';
import Projects from '../../sections/Projects';
import ScrollSection from './ScrollSection';
import Seabed from '../../sections/Seabed';
import Skills from '../../sections/Skills';
import OtherSkillsInfo from '../../sections/OtherSkills/OtherSkillsInfo';
import {
  ACTIVE_SECTION,
  BACK_ACTION,
  CUEPOINTS,
  EXPERIENCES,
  EXPERIENCES_SECTION,
  FORMATION_SECTION,
  INFO_PAGE_SECTION,
  OTHER_INFO_SECTION,
  PROJECTS,
  PROJECTS_SECTION,
  SEABED_AREA,
  SEABED_SECTION,
  SHOW_INSTRUCTIONS,
  SKILLS_FIRST_ROW,
  SKILLS_SECOND_ROW,
  SKILLS_SECTION,
  SKILLS_THIRD_ROW
} from '../../../constants';
import {
  ArrowIcon,
  LineOfArrows,
  MoreText,
  ScrollDownDisplay
} from '../../sections/MyInfoPage/styles';
import { ScrollAreaWrapper } from './styles';

const SCROLL_CUEPOINTS: Map<number, string> = new Map();
let touchEvent = 0;

declare global {
  interface Window {
    _debug: boolean;
  }
}

interface Props {
  isSeabedElementOpened: boolean;
  langModalVisible: boolean;
  openSeabedElement: Dispatch<SetStateAction<boolean>>;
}

const ScrollArea = ({
  isSeabedElementOpened,
  langModalVisible,
  openSeabedElement
}: Props): ReactElement => {
  const experiences: IExperiencesType[] = useSelector(experiencesSelector);
  const sectionsTexts: ISectionsText = useSelector(sectionsTextsSelector);
  const currentSection: string = useSelector(currentSecionSelector);

  const [cuePointsActivated, updateCuepoints] = useStateWithLabel<Set<string>>(
    new Set(),
    CUEPOINTS
  );
  const [activeSection, changeActiveSection] = useStateWithLabel<string>(
    INFO_PAGE_SECTION,
    ACTIVE_SECTION
  );
  const [showInstructions, toggleInstructions] = useStateWithLabel<boolean>(
    true,
    SHOW_INSTRUCTIONS
  );

  const dispatch = useDispatch();
  const scrollAreaRef = useRef<HTMLDivElement>(null);
  const touchActive = useRef<boolean>(false);

  const SECTIONS_INTERVAL_POINTS = {
    skills: {
      scrollAreaStart: 200,
      scrollAreaEnd: 1200
    },
    projects: {
      scrollAreaStart: 1000,
      scrollAreaEnd: 2000
    },
    experiences: {
      scrollAreaStart: 1400,
      scrollAreaEnd: 2100
    },
    seabed: {
      scrollAreaStart: 2600
    }
  };

  const experiencesSectionIds = new Set();

  const handleScroll = (event: React.UIEvent<HTMLDivElement, UIEvent>) => {
    const target = event.target as HTMLElement;
    const scrollPosition = target.scrollTop;
    const cuePointsAfterScrolling = new Set(cuePointsActivated);

    // eslint-disable-next-line no-underscore-dangle
    if (window._debug) {
      console.debug('scrollPosition', scrollPosition); // eslint-disable-line no-console
      console.debug('cuePointsActivated', cuePointsActivated); // eslint-disable-line no-console
    }

    SCROLL_CUEPOINTS.forEach((cuePointName, cuePoint) => {
      if (scrollPosition < cuePoint) {
        cuePointsAfterScrolling.delete(cuePointName);
      } else if (scrollPosition >= Number(cuePoint)) {
        if (cuePointName === EXPERIENCES) {
          // Makes experiences also go out when scrolling down
          //     experiencesSectionIds.forEach(experienceCuePoint => {
          //       cuePointsAfterScrolling.delete(experienceCuePoint);
        } else {
          cuePointsAfterScrolling.add(cuePointName);
        }
      }
    });
    // Prevents updating if cuepoints haven't changed
    if (cuePointsActivated.size !== cuePointsAfterScrolling.size) {
      updateCuepoints(new Set(cuePointsAfterScrolling));
    }
  };

  const resetScrollPosition = (delay: number): void => {
    const scrollArea = scrollAreaRef.current as HTMLDivElement;
    setTimeout(() => scrollArea.scrollTo(0, 0), delay);
  };

  const goToNextSection = useCallback(
    (type?: string): void => {
      if (type === BACK_ACTION) {
        dispatch(changeSection(getPreviousSection(currentSection)));
        changeActiveSection(getPreviousSection(currentSection));
      } else {
        dispatch(changeSection(getNextSection(currentSection)));
        changeActiveSection(getNextSection(currentSection));
      }
    },
    [dispatch, currentSection, changeActiveSection]
  );

  const handleTouchStart = useCallback((event: TouchEvent): void => {
    const xCoord: number = event.changedTouches[0].screenX;
    touchEvent = xCoord;
  }, []);

  const handleTouchEnd = useCallback(
    (event: TouchEvent): void => {
      const xCoord: number = event.changedTouches[0].screenX;
      const difference: number = touchEvent - xCoord;

      if (difference >= 100) {
        if (showInstructions && !langModalVisible) {
          event.stopPropagation();
          toggleInstructions(false);
          return;
        }
        goToNextSection();
      } else if (difference <= -100) {
        if (showInstructions && !langModalVisible) {
          event.stopPropagation();
          toggleInstructions(false);
          return;
        }
        goToNextSection(BACK_ACTION);
      }
      setTimeout(() => {
        touchEvent = 0;
      }, 1000);
    },
    [showInstructions, langModalVisible, goToNextSection, toggleInstructions]
  );

  const getNextSectionName = (): string =>
    sectionsTexts[getNextSection(currentSection)];

  useEffect(() => {
    const calculateSkillsScroll = (): void => {
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart,
        SKILLS_FIRST_ROW
      );
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart + 300,
        SKILLS_SECOND_ROW
      );

      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.skills.scrollAreaStart + 400,
        SKILLS_THIRD_ROW
      );
    };

    const calculateExperiencesScroll = (): void => {
      const NUMBER_OF_EXPERIENCES_DISPLAYED: number = experiences.length;
      const {
        scrollAreaStart,
        scrollAreaEnd
      }: {
        scrollAreaStart: number;
        scrollAreaEnd: number;
      } = SECTIONS_INTERVAL_POINTS.experiences;
      const intervalWindowForEachExperience: number =
        (scrollAreaEnd - scrollAreaStart) / NUMBER_OF_EXPERIENCES_DISPLAYED;
      let index = 0;

      for (
        let position = scrollAreaStart;
        position < scrollAreaEnd;
        position += intervalWindowForEachExperience
      ) {
        SCROLL_CUEPOINTS.set(Math.floor(position), experiences[index].id);
        experiencesSectionIds.add(experiences[index].id);
        index += 1;
      }
      SCROLL_CUEPOINTS.set(scrollAreaEnd, EXPERIENCES);
    };

    const calculateProjectsScroll = () => {
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.projects.scrollAreaStart,
        PROJECTS
      );
    };

    const calculateSeabedScroll = () => {
      SCROLL_CUEPOINTS.set(
        SECTIONS_INTERVAL_POINTS.seabed.scrollAreaStart,
        SEABED_AREA
      );
    };

    calculateSkillsScroll();
    calculateProjectsScroll();
    calculateExperiencesScroll();
    calculateSeabedScroll();
  }, [SECTIONS_INTERVAL_POINTS, experiences, experiencesSectionIds]);

  useEffect(() => {
    if (!isDesktop && activeSection === SKILLS_SECTION) {
      const allCuePoints: string[] = [
        SKILLS_FIRST_ROW,
        SKILLS_SECOND_ROW,
        SKILLS_THIRD_ROW
      ];
      updateCuepoints(new Set(allCuePoints));
    } else if (!isDesktop && activeSection === EXPERIENCES_SECTION) {
      const allCuePoints: string[] = [];
      let currentExperience = 0;

      const addNextExperience = (): void => {
        if (currentExperience === experiences.length - 1) {
          return;
        }
        allCuePoints.push(experiences[currentExperience].id);
        updateCuepoints(new Set(allCuePoints));
        currentExperience += 1;

        setTimeout(() => {
          addNextExperience();
        }, 50);
      };
      addNextExperience();
    }
  }, [activeSection, experiences, updateCuepoints]);

  useEffect(() => {
    if (!isDesktop && !touchActive.current) {
      document.addEventListener('touchstart', handleTouchStart);
      document.addEventListener('touchend', handleTouchEnd);
      touchActive.current = true;
    }
    return () => {
      document.removeEventListener('touchstart', handleTouchStart);
      document.removeEventListener('touchend', handleTouchEnd);
      touchActive.current = false;
    };
  }, [handleTouchStart, handleTouchEnd]);

  return (
    <ScrollAreaWrapper
      isSeabedElementOpened={isSeabedElementOpened}
      langModalVisible={langModalVisible}
      onScroll={handleScroll}
      ref={scrollAreaRef}
    >
      {!isDesktop && showInstructions && (
        <SwipeInstructions onCloseInstructions={toggleInstructions} />
      )}

      {(isDesktop || activeSection === INFO_PAGE_SECTION) && <MyInfoPage />}

      {(isDesktop || activeSection === SKILLS_SECTION) && (
        <ScrollSection title={sectionsTexts.technical}>
          <Skills cuePointsActivated={cuePointsActivated} />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === PROJECTS_SECTION) && (
        <ScrollSection title={sectionsTexts.projects}>
          <Projects cuePointsActivated={cuePointsActivated} />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === EXPERIENCES_SECTION) && (
        <ScrollSection title={sectionsTexts.experience}>
          <Experiences cuePointsActivated={cuePointsActivated} />
        </ScrollSection>
      )}

      {(isDesktop || activeSection === SEABED_SECTION) && (
        <ScrollSection title={sectionsTexts.other}>
          <Seabed
            cuePointsActivated={cuePointsActivated}
            openSeabedElement={openSeabedElement}
            resetScrollPosition={resetScrollPosition}
          />
        </ScrollSection>
      )}

      {!isDesktop && activeSection === FORMATION_SECTION && (
        <ScrollSection title={sectionsTexts.other}>
          <Formation status={{ active: true, read: true, visible: true }} />
        </ScrollSection>
      )}

      {!isDesktop && activeSection === OTHER_INFO_SECTION && (
        <ScrollSection title={sectionsTexts.other}>
          <OtherSkillsInfo visible />
        </ScrollSection>
      )}

      {!isDesktop && (
        <ScrollDownDisplay sections>
          <MoreText onClick={() => goToNextSection()}>
            {getNextSectionName()}
          </MoreText>
          <LineOfArrows>
            <ArrowIcon
              className="fas fa-angle-double-down"
              goBack={activeSection === OTHER_INFO_SECTION}
              sections
            />
          </LineOfArrows>
        </ScrollDownDisplay>
      )}
    </ScrollAreaWrapper>
  );
};

ScrollArea.propTypes = {
  langModalVisible: bool.isRequired,
  isSeabedElementOpened: bool.isRequired,
  openSeabedElement: func.isRequired
};

export default ScrollArea;
