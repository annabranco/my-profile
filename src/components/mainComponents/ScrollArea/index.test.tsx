import { shallow, ShallowWrapper } from 'enzyme';
import 'jest-styled-components';
import '../../../../testing/mocks/matchMedia';
import ScrollArea from '.';
import { setupMockProvider } from '../../../../testing/mocks';

// TODO [10-May -21]: Implement defined testing suite (Anna Branco)

describe('< ScrollArea >', () => {
  let MockProvider;
  let wrapper: ShallowWrapper;

  beforeAll(() => {
    MockProvider = setupMockProvider({ language: 'en' }).MockProvider;
    wrapper = shallow(
      <MockProvider>
        <ScrollArea
          isSeabedElementOpened={false}
          langModalVisible={false}
          openSeabedElement={() => jest.fn()}
        />
      </MockProvider>
    );
  });

  it('should mount', () => {
    expect(wrapper).toBeDefined();
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  xit('should display navigation instructions when mounting on devices', () => {
  });

  xit('should change the active section when scrolling', () => {
  });

  xit('should display the corresponding component when its section is active on desktops', () => {
  });

  xit('should display the corresponding component when its section is active on devices', () => {
  });

  xit('should display a button that advances to the next section on devices', () => {
  });

  xit('should advance to the next section when swiped right on devices', () => {
  });

  xit('should back to the previous section when swiped left on devices', () => {
  });
});
