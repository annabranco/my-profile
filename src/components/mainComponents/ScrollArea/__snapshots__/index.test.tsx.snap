// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< ScrollArea > matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ScrollArea
    isSeabedElementOpened={false}
    langModalVisible={false}
    openSeabedElement={[Function]}
  />
</Provider>
`;
